Settings - Persist Payment Method Checkbox After Capability Request

htdat opened the pull request:
Fixes WOOPMNT-4963

#### Changes proposed in this Pull Request

- Move the check of active payment method from the Settings Controller method in `update_enabled_payment_methods` to the `gateway->is_available`
- This change allows the merchant to save selected payment methods when these payment methods require additional reviews, i.e. in the pending state for a while. This applies to Alipay and Wechat Pay. 
- When the payment method capability changes from `pending` to `active`, it keeps being selected and can display in the checkout. Before the capability switches to `active`, it will be filtered out from the checkout page.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Enable WeChat Pay and Alipay via the metadata in your local server. Needs to use a US-based account. 
- In WP Admin > WooPayments > Settings, try to enable WeChat Pay or Alipay or both. 
- Expect: the selected payment method shows Pending Verification.
- Expect: the checkout page does not display WeChat Pay or Alipay. 
- Go to WP Admin > WooPayments > Overview, update more info to remove the Restricted status and more info.
- Expect: WP Admin > WooPayments > Settings, WeChat Pay / Alipay are still checked.
- Expect: the checkout page displays WeChat Pay / Alipay

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

htdat committed:
Move the check of active payment method to gateway->is_available

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10754` or branch name `update/persist-wechat-pay-alipay-checked-state` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/persist-wechat-pay-alipay-checked-state) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 75004634fda9ba3e334d2ccd02169eb92f198ab2
- Build time: 2025-05-08 10:32:26 UTC

_Note: the build is updated when a new commit is pushed to this PR._

htdat committed:
Merge branch 'develop' into update/persist-wechat-pay-alipay-checked-state

htdat committed:
Remove request-capability endpoint

- Added here 281c2a00eac31e9a041f90721eaaddb9e0548790
- Most of the code in the commit above was removed in dcca5d85ebce9b8b4d25b8ab6fabafc6ef82ec85 and 05109176c7b3ead20286f5b9fe1a0b592fdc712a

htdat committed:
Fix + add tests

htdat commented on includes/admin/class-wc-rest-payments-settings-controller.php:
```diff
@@ -310,15 +310,6 @@ public function register_routes() {
 				'permission_callback' => [ $this, 'check_permission' ],
 			]
 		);
-		register_rest_route(
-			$this->namespace,
-			'/' . $this->rest_base . '/request-capability',
```
I've noticed this endpoint is now redundant. A bit of history:

- Added here 281c2a00eac31e9a041f90721eaaddb9e0548790 to have JCB request capability notice. 
- Most of the code in the commit above was removed in dcca5d85ebce9b8b4d25b8ab6fabafc6ef82ec85 and 05109176c7b3ead20286f5b9fe1a0b592fdc712a

htdat committed:
changelog

frosso reviewed (approved):
Tested locally. All good. I like the addition of the tests.

Confirmed that rejected/approved/pending payment methods are correctly added to (or excluded from) checkout.
Confirmed that the payment method selection is persisted even if the payment method becomes "pending".

<img width="1050" alt="Screenshot 2025-05-08 at 12 55 50 PM" src="https://github.com/user-attachments/assets/42207990-b4c1-4167-b652-ac5df4132711" />

<img width="1304" alt="Screenshot 2025-05-08 at 1 10 04 PM" src="https://github.com/user-attachments/assets/bb2d5906-a331-4c2a-8b9c-c227e28903f3" />
<img width="1416" alt="Screenshot 2025-05-08 at 1 09 52 PM" src="https://github.com/user-attachments/assets/0648c5f4-a52d-42c0-9708-35abd9f5432b" />
